### ISSUE TITLE
When Consul goes beyond the max number of files descriptors, it cannot recover

### ISSUE THREAD

**Issue by pierresouchay** [hashicorp/consul#3886]
When Consul goes out of file descriptors (in our case, limited by Systemd), it cannot recover from its state and is indefinitely not working well.

We had this issue on several agents (were the number of FDs did overcome the 8k limit we put on file descriptors) and the agents were then unable to recover from their state (for instance, all checks were returning errors saying Consul could not create file descriptors anymore), even when the load on Consul agent was reduced (here, too many processes performing too many blocking queries)

Only a full restart of agent did succeeded into make it work again

**Comment by preetapan** [comment]
@pierresouchay the next time you see this happen, could you run `lsof -p` on the agent's PID and post the results here. Would be useful for us to track down where the leaky file handles are coming from.

**Comment by pierresouchay** [comment]
It is trivial to reproduce:
```
 ulimit -n 256 
```

Launch consul ...
watch a few endpoints (for instance using consul-template)...

I can do it if you want

**Comment by pierresouchay** [comment]
Ok, to be fair, it is not that easy to reproduce:

I set ulimit to a low value.
I added lots of watches several times on the agent, I had messages like this:
```
[WARN] agent: http request failed 'http://localhost:31643/': Get http://localhost:31643/: dial tcp 127.0.0.1:31643: socket: too many open files
```
But when I did stop the watches, the messages do go away.

I saw this message several times on several nodes, but not sure whether if it was on the same version of Consul, was not able to reproduce.

I might be interesting however to set a quota of FD to the consul agent in order to avoid being DOSed by clients.

Such as (max_number_of_fds - connections_to_consul_servers - max_gossip_connections - max_number_of_checks)

Might be difficult to implement however, you might close this issue if you think it is a corner case.

Regards
